pipelines/common/src/main/java/com/google/fhir/analytics/view/ViewDefinition.java

```diff
@@ -15,44 +15,169 @@
+   * @return the set of new column names.
+   * @throws ViewDefinitionException for repeated columns or other requirements not satisfied.
+   */
+  private Set<String> validateAndReplaceConstantsInSelects(
+      List<Select> selects, Set<String> currentColumns) throws ViewDefinitionException {
+    Set<String> newCols = new HashSet<>();
+    if (selects == null) {
+      return newCols;
+    }
+    for (Select s : selects) {
+      newCols.addAll(
+          validateAndReplaceConstantsInOneSelect(s, Sets.union(currentColumns, newCols)));
+    }
+    return newCols;
   }
 
-  public boolean validate() {
-    // TODO: Add checks like resource type/version validation, etc.
-    return true;
+  private Set<String> validateAndReplaceConstantsInOneSelect(
+      Select select, Set<String> currentColumns) throws ViewDefinitionException {
+    Set<String> newCols = new HashSet<>();
+    if (select.getColumn() != null) {
+      for (Column c : select.getColumn()) {
+        if (Strings.nullToEmpty(c.name).isEmpty()) {
+          throw new ViewDefinitionException("Column name cannot be empty!");
+        }
+        if (Strings.nullToEmpty(c.path).isEmpty()) {
+          throw new ViewDefinitionException("Column path cannot be empty for " + c.name);
+        }
+        if (currentColumns.contains(c.getName()) || newCols.contains(c.getName())) {
+          throw new ViewDefinitionException("Repeated column name " + c.getName());
+        }
+        newCols.add(c.getName());
+        c.path = validateAndReplaceConstants(c.getPath());
+      }
+    }
+    if (!Strings.nullToEmpty(select.getForEach()).isEmpty()) {
+      select.forEach = validateAndReplaceConstants(select.getForEach());
+    }
+    if (!Strings.nullToEmpty(select.getForEachOrNull()).isEmpty()) {
+      select.forEachOrNull = validateAndReplaceConstants(select.getForEachOrNull());
+    }
+    newCols.addAll(
+        validateAndReplaceConstantsInSelects(
+            select.getSelect(), Sets.union(currentColumns, newCols)));
+    Set<String> unionCols = null;
+    if (select.getUnionAll() != null) {
+      for (Select u : select.getUnionAll()) {
+        Set<String> uCols =
+            validateAndReplaceConstantsInOneSelect(u, Sets.union(currentColumns, newCols));
+        if (unionCols == null) {
+          unionCols = uCols;
+        } else {
+          if (!unionCols.equals(uCols)) {
+            throw new ViewDefinitionException(
+                "Union columns are not consistent "
+                    + Arrays.toString(uCols.toArray())
+                    + " vs "
+                    + Arrays.toString(unionCols.toArray()));
+          }
+        }
+      }
+    }
+    if (unionCols != null) {
+      return Sets.union(newCols, unionCols);
+    }
+    return newCols;
+  }
+
+  private String validateAndReplaceConstants(String fhirPath) throws ViewDefinitionException {
+    Matcher matcher = CONSTANT_PATTERN.matcher(fhirPath);
```

Comment:
Not sure if this helps but was checking [this](https://github.com/hapifhir/hapi-fhir/blob/e15d0430d0aaac352f4e574773f84375129ef42e/hapi-fhir-base/src/main/java/ca/uhn/fhir/fhirpath/IFhirPath.java#L88) which potentially can be used

Reply:
Yeah that's similar to the issue I am currently dealing with for implementing type support (since it is needed for DB export). I don't think `parse` alone is enough because the context also matters (e.g., the `IBase` in `evaluate`). For now I am postponing that (you will see it in the next PR soon) and instead am planning to file a bug for it to do later. The main reason is that it is not trivial AFAICT.

Reply:
Sure